_posts/2025-05-09-sorting-for-fast-selective-queries.md

```diff
@@ -0,0 +1,660 @@
+Our advanced techniques are nearly as fast as the dedicated sort on `origin`.
+
+<div id="remote_s3_query_performance_by_destination" style="width:100%;height:400px;min-width:720px;"></div>
+<script>
+    fetch('{{ site.baseurl }}/data/zonemaps/remote_s3_query_performance_by_destination.json')
+        .then(res => res.json())
+        .then(parsed_json => {
+            let my_element = document.getElementById('remote_s3_query_performance_by_destination');
+            Plotly.newPlot( my_element, parsed_json.data, parsed_json.layout );
+            });
+</script>
+
+However, if you look at the plot **by `destination`**, we see the value of the more balanced techniques.
+The `origin_dest` approach of just appending `destination` to the list of sorted columns misses out on much of the benefits.
+
+<div id="remote_s3_query_performance_by_origin_destination" style="width:100%;height:400px;min-width:720px;"></div>
+<script>
+    fetch('{{ site.baseurl }}/data/zonemaps/remote_s3_query_performance_by_origin_destination.json')
+        .then(res => res.json())
+        .then(parsed_json => {
+            let my_element = document.getElementById('remote_s3_query_performance_by_origin_destination');
+            Plotly.newPlot( my_element, parsed_json.data, parsed_json.layout );
+            });
+</script>
+
+When querying for a **specific `origin` and `destination` pair,** all approaches are significantly faster than random.
+However, zooming in on the non-random techniques, the more advanced sorting approaches are as fast or slightly faster than sorting by `origin` or by `origin` and then `destination`.
+Performance is also faster than the other experiments, as less data needs to be read (since the filter is more selective).
+
+### Approximate Time Sorting
+
+More recent data tends to be more useful data, so frequently queries filter on a time column.
+However, often queries filter on time and on other columns as well.
+
+**Don't just sort on the timestamp column!**
+You will miss out on performance benefits.
+This is because timestamps tend to be so granular, that in practical terms the data is only sorted by timestamp.
+How many rows of your data were inserted at exactly `2025-01-01 01:02:03.456789`?
+Probably just one!
+
+To sort on multiple columns as well as a time column, first sort by a truncated timestamp and then on the other columns.
+In this experiment, we truncate the `flightdate` column to three levels of granularity: day, month, and year.
+We then use our most effective multi-column approach and sort by Hilbert encoding of `origin` and `dest` next.
+
+For each of the query patterns tested previously (filters on `origin`, `destination`, and `origin` / `destination`), we filter on 3 time ranges: the latest 1 week, 13 weeks, and 52 weeks.
+This yields a total of 9 scenarios.
```

Comment:
It's a total of 3x3x3 = 27 scenarios. Part of the separation happens among the plots, another part happens within the plots. Please elaborate on this.

Reply:
Good catch! Yes, the experiment needs some more explanation.

Reply:
This will be in the follow up post! Thanks for the feedback!